validator/src/main/java/com/amazon/aoc/validators/LoadBalanceValidator.java

```diff
@@ -0,0 +1,170 @@
+  }
+
+  @Override
+  public void init(
+      Context context, ValidationConfig validationConfig, ICaller caller, FileConfig expectedTrace)
+      throws Exception {
+    this.xrayService = new XRayService(context.getRegion());
+    this.caller = caller;
+    this.context = context;
+  }
+
+  @Override
+  public void validate() throws Exception {
+    // 2 retries for calling the sample app to handle the Lambda case,
+    // where first request might be a cold start and have an additional unexpected subsegment
+    int successes = 0;
+    while (successes < 5) {
+      // Call sample app and get locally stored trace
+      String traceId = this.getAppTraceId();
+      List<String> traceIdList = Collections.singletonList(traceId);
+      ;
+      log.info("value of Sample App traceId: {}", traceId);
+
+      // Retry 5 times to since segments might not be immediately available in X-Ray service
+      RetryHelper.retry(
+          5,
+          Integer.parseInt(GenericConstants.SLEEP_IN_MILLISECONDS.getVal()),
+          true,
+          () -> {
+            // get retrieved trace from x-ray service
+            Map<String, Object> retrievedTrace = this.getRetrievedTrace(traceIdList);
+            log.info("value of retrieved trace map: {}", retrievedTrace);
+            Set<String> set =
+                retrievedTrace.keySet().stream()
+                    .filter(s -> s.endsWith("collector-id"))
+                    .collect(Collectors.toSet());
+            if (set.size() < 2) {
+              log.error("not enough spans in trace map (need at least 2)");
+              log.info("==========================================");
+              throw new BaseException(ExceptionCode.NOT_ENOUGH_SPANS);
+            }
+            String targetCollectorId = null;
+            for (String collectorId : set) {
+              if (targetCollectorId == null) {
+                targetCollectorId = retrievedTrace.get(collectorId).toString();
+              } else {
+                if (!targetCollectorId.equals(retrievedTrace.get(collectorId).toString())) {
+                  log.info("id values do not match");
+                  log.info("value of target id: {}", targetCollectorId);
+                  log.info("value of retrieved id: {}", retrievedTrace.get(collectorId).toString());
+                  log.info("==========================================");
+                  throw new BaseException(ExceptionCode.COLLECTOR_ID_NOT_MATCHED);
+                }
+              }
+            }
+          });
+
+      successes += 1;
+      log.info("Total number of successful runs: {}", successes);
+    }
+
+    log.info("validation is passed for path {}", caller.getCallingPath());
+  }
+
+  // this method will hit get trace from x-ray service and get retrieved trace
+  private Map<String, Object> getRetrievedTrace(List<String> traceIdList) throws Exception {
+    List<Trace> retrieveTraceList = xrayService.listTraceByIds(traceIdList);
```

Comment:
Does XRayService allow us to query by a single trace id instead of a collection of them?

Reply:
I have found that we store it in a singleton list because this is the [method](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/xray/model/BatchGetTracesRequest.html#:~:text=was%20not%20specified.-,traceIds,-public%20final%C2%A0) required to get the segments. It takes in a list of traceIds.